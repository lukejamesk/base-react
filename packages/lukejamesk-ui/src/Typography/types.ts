export type TypographyIntrinsicElements = Required<
  Pick<
    JSX.IntrinsicElements,
    'p' | 'strong' | 'blockquote' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span'
  >
>

export type TypographyVariant = Required<
  | 'paragraph'
  | 'bold'
  | 'blockquote'
  | 'heading1'
  | 'heading2'
  | 'heading3'
  | 'heading4'
  | 'heading5'
  | 'heading6'
  | 'text'
>
