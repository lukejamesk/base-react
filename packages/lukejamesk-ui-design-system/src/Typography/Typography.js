import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import { Typography as TypographyCore } from '@lukejamesk/core'

const variantMapping = {
  paragraph: 'ljk-typo-p',
  bold: 'ljk-typo-b',
  blockquote: 'ljk-typo-blockquote',
  heading1: 'ljk-typo-h1',
  heading2: 'ljk-typo-h2',
  heading3: 'ljk-typo-h3',
  text: 'ljk-typo-text',
}

export const typographyVariants = [
  'paragraph',
  'bold',
  'blockquote',
  'heading1',
  'heading2',
  'heading3',
  'text',
]

const Typography = forwardRef(({ variant = 'text', className, paragraph = false, ...other }, ref) => {
  const v = paragraph ? variantMapping.paragraph : variantMapping[variant]

  return <TypographyCore variant={variant} className={clsx(v, className)} ref={ref} {...other} />
})

Typography.displayName = 'Typography'

Typography.propTypes = {
  variant: PropTypes.oneOf(typographyVariants),
  paragraph: PropTypes.bool,
  className: PropTypes.string,
}

export default Typography
