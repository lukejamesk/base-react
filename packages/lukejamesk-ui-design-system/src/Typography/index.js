export { default, typographyVariants } from './Typography'
