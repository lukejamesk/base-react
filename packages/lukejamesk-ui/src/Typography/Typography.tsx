import React, { forwardRef } from 'react'
import { TypographyVariant, TypographyIntrinsicElements } from './types'

export const typographyVariants: Record<TypographyVariant, keyof TypographyIntrinsicElements> = {
  paragraph: 'p',
  bold: 'strong',
  blockquote: 'blockquote',
  heading1: 'h1',
  heading2: 'h2',
  heading3: 'h3',
  heading4: 'h4',
  heading5: 'h5',
  heading6: 'h6',
  text: 'span',
}

export type TypographyProps = React.PropsWithChildren<{
  variant?: TypographyVariant
  className?: string | undefined
}>

export type Ref = any

const findVariant = (variant?: TypographyVariant) => {
  return variant ? typographyVariants[variant] : typographyVariants.paragraph
}

const Typography = forwardRef<Ref, TypographyProps>(({ variant, className, ...other }, ref) => {
  const Component: keyof TypographyIntrinsicElements = findVariant(variant)
  return <Component className={className} {...other} ref={ref} />
})

Typography.displayName = 'Typography'

export default Typography
