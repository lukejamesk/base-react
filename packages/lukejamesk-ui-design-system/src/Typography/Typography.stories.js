import React from 'react'
import { text, boolean } from '@storybook/addon-knobs'
import Typography from './Typography'

const exampleText = `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Maecenas feugiat varius mi, eu sagittis nunc dignissim nec.
Quisque rhoncus hendrerit leo.`

export default {
  title: 'Design System|Typography',
  component: Typography,
}

export const Paragraph = () => <Typography paragraph={boolean}>{text('text', exampleText)}</Typography>
export const Blockquote = () => <Typography variant="blockquote">{text('text', exampleText)}</Typography>
export const Bold = () => <Typography variant="bold">{text('text', exampleText)}</Typography>
export const Heading1 = () => <Typography variant="heading1">{text('text', 'Heading 1')}</Typography>
export const Heading2 = () => <Typography variant="heading2">{text('text', 'Heading 2')}</Typography>
export const Heading3 = () => <Typography variant="heading3">{text('text', 'Heading 3')}</Typography>
