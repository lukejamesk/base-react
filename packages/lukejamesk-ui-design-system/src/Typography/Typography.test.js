import React from 'react'
import { render } from 'enzyme'
import { forEach } from 'ramda'
import Typography, { typographyVariants } from './Typography'

describe('<Typography />', () => {
  forEach((variant) => {
    describe(`Variant: ${variant}`, () => {
      it('should render', () => {
        const wrapper = render(<Typography variant={variant}>My content</Typography>)
        expect(wrapper).toMatchSnapshot()
      })
    })
  }, typographyVariants)
})
