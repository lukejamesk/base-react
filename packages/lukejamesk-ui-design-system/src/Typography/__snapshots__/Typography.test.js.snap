// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> Variant: blockquote should render 1`] = `
<blockquote
  class="ljk-typo-blockquote"
>
  My content
</blockquote>
`;

exports[`<Typography /> Variant: bold should render 1`] = `
<strong
  class="ljk-typo-b"
>
  My content
</strong>
`;

exports[`<Typography /> Variant: heading1 should render 1`] = `
<h1
  class="ljk-typo-h1"
>
  My content
</h1>
`;

exports[`<Typography /> Variant: heading2 should render 1`] = `
<h2
  class="ljk-typo-h2"
>
  My content
</h2>
`;

exports[`<Typography /> Variant: heading3 should render 1`] = `
<h3
  class="ljk-typo-h3"
>
  My content
</h3>
`;

exports[`<Typography /> Variant: paragraph should render 1`] = `
<p
  class="ljk-typo-p"
>
  My content
</p>
`;

exports[`<Typography /> Variant: text should render 1`] = `
<span
  class="ljk-typo-text"
>
  My content
</span>
`;
